diff -ur veewee-0.4.5.1.orig/lib/veewee/provider/kvm/provider.rb veewee-0.4.5.1/lib/veewee/provider/kvm/provider.rb
--- veewee-0.4.5.1.orig/lib/veewee/provider/kvm/provider.rb	2014-12-28 20:10:19.000000000 +0900
+++ veewee-0.4.5.1/lib/veewee/provider/kvm/provider.rb	2014-12-04 16:19:40.000000000 +0900
@@ -12,7 +12,7 @@
           Fog.credentials[:libvirt_uri] ||= "qemu:///system"
 
           env.logger.info "Setting libvirt IP Command if not already defined in .fog config file"
-          Fog.credentials[:libvirt_ip_command] ||= "arp -an |grep -i $mac|cut -d '(' -f 2 | cut -d ')' -f 1"
+          Fog.credentials[:libvirt_ip_command] ||= "/sbin/arp -an |grep -i $mac|cut -d '(' -f 2 | cut -d ')' -f 1"
 
           env.logger.info "Opening a libvirt connection using fog.io"
           begin
@@ -46,7 +46,7 @@
           conn.pools.any? or raise Veewee::Error, "You need at least one (active) storage pool defined in #{Fog.credentials[:libvirt_uri]}."
 
           env.logger.info "Checking availability of the arp utility"
-          shell_exec("arp -an").status.zero? or raise Veewee::Error, "Could not execute the arp command. This is required to find the IP address of the VM"
+          shell_exec("/sbin/arp -an").status.zero? or raise Veewee::Error, "Could not execute the arp command. This is required to find the IP address of the VM"
 
         end
 
